File: adapters/database/src/main/java/net/onelitefeather/pandorascluster/database/mapper/impl/LandMemberMapper.java
package net.onelitefeather.pandorascluster.database.mapper.impl;

import net.onelitefeather.pandorascluster.api.mapper.DatabaseEntityMapper;
import net.onelitefeather.pandorascluster.api.player.LandMember;
import net.onelitefeather.pandorascluster.database.models.player.LandMemberEntity;
import net.onelitefeather.pandorascluster.database.models.player.LandPlayerEntity;
import net.onelitefeather.pandorascluster.dbo.player.LandMemberDBO;
import org.jetbrains.annotations.Nullable;

public final class LandMemberMapper implements DatabaseEntityMapper<LandMemberDBO, LandMember> {

    private final LandPlayerMapper landPlayerMapper;

    public LandMemberMapper(LandPlayerMapper landPlayerMapper) {
        this.landPlayerMapper = landPlayerMapper;
    }

    @Override
    public @Nullable LandMember entityToModel(@Nullable LandMemberDBO entity) {
        if (entity == null) return null;
        var landPlayer = landPlayerMapper.entityToModel(entity.member());
        if (landPlayer == null) return null;
        return new LandMember(entity.id(), landPlayer, entity.role());
    }

    @Override
    public @Nullable LandMemberDBO modelToEntity(@Nullable LandMember model) {
        if (model == null) return null;
        LandPlayerEntity landPlayer = (LandPlayerEntity) landPlayerMapper.modelToEntity(model.getMember());
        if (landPlayer == null) return null;
        return new LandMemberEntity(model.getId(), landPlayer, model.getRole(), null);
    }
}

File: adapters/database/src/main/kotlin/net/onelitefeather/pandorascluster/database/mapper/impl/LandMemberMapper.kt
package net.onelitefeather.pandorascluster.database.mapper.impl

import net.onelitefeather.pandorascluster.api.mapper.DatabaseEntityMapper
import net.onelitefeather.pandorascluster.api.player.LandMember
import net.onelitefeather.pandorascluster.api.service.DatabaseService
import net.onelitefeather.pandorascluster.database.models.player.LandMemberEntity
import net.onelitefeather.pandorascluster.database.models.player.LandPlayerEntity
import net.onelitefeather.pandorascluster.dbo.player.LandMemberDBO

class LandMemberMapper(private val databaseService: DatabaseService) :
    DatabaseEntityMapper<LandMemberDBO, LandMember> {

    override fun entityToModel(entity: LandMemberDBO?): LandMember? {
        if (entity == null) return null

        val member = databaseService.landPlayerMapper().entityToModel(entity.member()) ?: return null
        return LandMember(entity.id(), member, entity.role())
    }

    override fun modelToEntity(model: LandMember?): LandMemberDBO? {
        if (model == null) return null
        val member = databaseService.landPlayerMapper().modelToEntity(model.member) ?: return null
        return LandMemberEntity(model.id, member as LandPlayerEntity, model.role)
    }
}